File: templates/go-aoc/internal/get-aoc-input/command.go
package aocinput

import (
	"fmt"
	"os"
)

const errorPref = "ERROR SCAFFOLDING AOC INPUT: "

type CLIInput struct {
	Output  string
	Year    int
	Day     int
	Session string
}

// Scaffold downloads the day's input and writes the task skeleton. Errors
// are printed here (not returned) since main is the only caller and has no
// use for a returned error beyond printing it anyway.
func Scaffold(in CLIInput) {
	fmt.Printf("SCAFFOLDING AOC INPUT FOR YEAR %d DAY %d...\n", in.Year, in.Day)

	if err := os.MkdirAll(in.Output, 0o755); err != nil {
		fmt.Fprintln(os.Stderr, errorPref+err.Error())
		return
	}

	payload := FetchRequestInput{Year: in.Year, Day: in.Day, Session: in.Session}

	input, err := DownloadInput(payload)
	if err != nil {
		fmt.Fprintln(os.Stderr, errorPref+err.Error())
		return
	}
	if err := WriteTask(WriteTaskInput{OutputDir: in.Output, Data: input}); err != nil {
		fmt.Fprintln(os.Stderr, errorPref+err.Error())
		return
	}

	title, err := FetchTitle(payload)
	if err != nil {
		fmt.Fprintln(os.Stderr, errorPref+err.Error())
		return
	}
	if err := WriteTemplate(WriteTemplateInput{OutputDir: in.Output, Year: in.Year, Day: in.Day, Title: title}); err != nil {
		fmt.Fprintln(os.Stderr, errorPref+err.Error())
		return
	}

	fmt.Printf("AOC INPUT FOR YEAR %d DAY %d SCAFFOLDED SUCCESSFULLY!\n", in.Year, in.Day)
}

File: templates/rust-aoc/lib/get_aoc_input/command.rs
use std::collections::HashMap;
use std::fs;
use std::process;

use garde::Validate;

use super::client;
use super::scaffold;

// Bounds live as tags on the struct so garde checks them together, rather                                                                ║
// than a scattered if per field.
#[derive(Validate)]
pub struct CliInput {
    #[garde(range(min = 1, max = 25))]
    pub day: u32,
    #[garde(length(min = 2))]
    pub session: String,
    #[garde(skip)]
    pub year: u32,
    #[garde(skip)]
    pub output: String,
}

pub fn validation_messages() -> HashMap<&'static str, &'static str> {
    HashMap::from([
        ("day", "Invalid day. Must be a number between 1 and 25."),
        (
            "session",
            "Invalid session key. Must be longer than 1 character.",
        ),
    ])
}

/// Downloads the day's input and writes the task skeleton. Errors are printed
/// here (not returned) since main is the only caller and has no use for a
/// returned error beyond printing it anyway.
pub fn scaffold(input: CliInput) {
    let CliInput {
        day,
        year,
        session,
        output,
    } = input;
    eprintln!("SCAFFOLDING AOC INPUT FOR YEAR {year} DAY {day}...");

    if let Err(e) = fs::create_dir_all(&output) {
        eprintln!("ERROR SCAFFOLDING AOC INPUT: {e}");
        process::exit(0);
    }

    match client::download_input(year, day, &session) {
        Ok(task_input) => {
            if let Err(e) = scaffold::write_task(scaffold::WriteTaskInput {
                output_dir: &output,
                data: &task_input,
            }) {
                eprintln!("ERROR SCAFFOLDING AOC INPUT: {e}");
                process::exit(0);
            }
        }
        Err(e) => {
            eprintln!("ERROR SCAFFOLDING AOC INPUT: {e}");
            process::exit(0);
        }
    }

    let title = client::fetch_title(year, day, &session);
    if let Err(e) = scaffold::write_template(scaffold::WriteTemplateInput {
        output_dir: &output,
        year,
        day,
        title: &title,
    }) {
        eprintln!("ERROR SCAFFOLDING AOC INPUT: {e}");
        process::exit(0);
    }

    if let Err(e) = scaffold::register_bin(day) {
        eprintln!("ERROR REGISTERING BINARY: {e}");
        process::exit(0);
    }

    eprintln!("AOC INPUT FOR YEAR {year} DAY {day} SCAFFOLDED SUCCESSFULLY!");
}
